lib/domain/usecase/room/chat_room_search_interactor.dart

```diff
@@ -0,0 +1,66 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/config/app_config.dart';
+import 'package:fluffychat/domain/app_state/room/chat_room_search_state.dart';
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:matrix/matrix.dart';
+
+class ChatRoomSearchInteractor {
+  Stream<Either<Failure, Success>> execute({
+    required Timeline timeline,
+    required String keyword,
+    required int currentIndex,
+    required bool isGoDown,
+    required int maxRequestHistory,
+    required int historyCount,
+  }) async* {
+    try {
+      if (keyword.length < AppConfig.chatRoomSearchKeywordMin) {
+        yield Right(ChatRoomSearchInitial());
+        return;
+      }
+      yield Right(ChatRoomSearchLoading());
+      final bool Function(Event)? searchFunc;
+      if (AppConfig.chatRoomSearchWordStrategy) {
+        searchFunc = (event) => event.plaintextBody.containsWord(keyword);
+      } else {
+        final lowerCaseKeyword = keyword.toLowerCase();
+        searchFunc = (event) =>
+            event.plaintextBody.toLowerCase().contains(lowerCaseKeyword);
+      }
+      if (isGoDown) {
+        final index =
+            timeline.events.sublist(0, currentIndex).lastIndexWhere(searchFunc);
+        if (index >= 0) {
+          yield Right(
+            ChatRoomSearchSuccess(keyword: keyword, eventIndex: index),
+          );
+          return;
+        }
+        yield Left(ChatRoomSearchNoResult());
+        return;
+      }
+      for (var i = 0; i < maxRequestHistory; i++) {
+        final index =
+            timeline.events.sublist(currentIndex + 1).indexWhere(searchFunc);
```

Comment:
this function don't depend on `i`, why put it inside the for loop

Reply:
srr, but not easy to understand it.

Reply:
Can you describe more details about `maxRequestHistory`?